gRPC core: strip zone-id from IPv6 hosts before TLS verification

dcow opened the pull request:
When initiating a connection to an IPv6 peer using an address that is
not globally scoped, there may be ambiguity regarding which zone the
destination address applies to when multiple links of the same scope
are present. The scoped address architecture and zone-id syntax are
described in rfc4007 and rfc 6874, respectively:

  * https://tools.ietf.org/html/rfc4007#section-6
  * https://tools.ietf.org/html/rfc6874

This patch allows host name verification performed during TLS session
establishment, and on a per-call basis, to work correctly when the peer
presents a certificate with a non-global IPv6 address listed as one of
its alternate names. Whether arbitrary certificate authorities choose
issue certificates of this nature, or not, is outside the scope of gRPC.

The zone-id is separated from the address using a percent (%) character.
It is considered a system implementation detail and guidance suggests it
be stripped from any paths or addresses egressing the host because it is
irrelevant and meaningless otherwise. Thus, it would not make sense for
a server to present a certificate containing non-global IPv6 addresses
with zone-ids present.

Resolves #14371

---

NOTE: I expect this will require a little cleanup and restructuring/testing
but I want to put it up to get discussion started. If needed, we could also
address #14369 at the same time.

markdroth commented:
Jiangtao and Julien should approve this as a change to the security code.  Also requesting review from Paul as our resident IPv6 expert.

dcow commented:
If this raises too many eyebrows, I think a more natural option for us could be to use ULAs. I put this together before I totally realized site-locals were replaced with ULAs. But, even if we go with ULAs as our working solution, maybe this could still be beneficial?

pmarks-net commented on src/core/lib/security/transport/security_connector.cc:
```diff
@@ -704,18 +704,31 @@ static void ssl_server_add_handshakers(grpc_server_security_connector* sc,
   }
+
+  /* IPv6 non-global zone-id should not be included in comparisons. */
+  const char* percent = strchr(peer_name, '%');
```
Could `%` possibly mean anything else in this context?  I'm thinking about cases like `google.com%2Eattacker.evil` or `evil.news%2Egoogle.com`.

Every IPv6 address contains a `:`, so you could avoid a malloc by truncating `allocated_host`.

In general, an IPv6 link-local address is almost as low-level as a MAC address: every address in the world shares the same /64.  Is it actually plausible that anyone would operate TLS at this layer in the stack?

dcow replied:
I work on a networking product. We'd like to be able to securely diagnose issues with network configuration when higher levels have potentially failed. So that's one use-case. I agree I can't think of many others off the top of my head. 

re `%` in domain names traditionally they're just alphanumeric and hyphen. But there's been some more recent push regarding support for unicode in hostnames https://en.wikipedia.org/wiki/Punycode and I don't know if gRPC a) supports this and b) would perform the conversion to plain ascii when checking alternate names.

dcow replied:
@pmarks-net re allocation: I didn't want to modify `allocated_name` from `grpc_split_host_port` but now that you mention it I can't think of anything bad about that approach and the fewer allocations the better. Thanks!

pmarks-net replied:
> We'd like to be able to securely diagnose issues with network configuration when higher levels (slaac, dhcp, dns, etc.) have potentially failed.

Will this change actually allow you to do that?  If I understand correctly, you would need to obtain TLS certificates for your link-local addresses, which sounds very strange to me. Does gRPC work well with self-signed certificates?

dcow replied:
> Does gRPC work well with self-signed certificates?

Yep! You can specify your own list of root trust anchors. We are self-issuing certificates during our provisioning stage and compiling the trust roots for use in our application layer (which uses gRPC). I have verified that with this fix I'm able to connect to a service on one of my hosts via its link-local address without explicitly overriding the expected host name.

dcow replied:
rgr I will update/simplify when I fix the conflicts. Stay tuned.

dcow commented:
Gentle ping to @jboeuf @jiangtaoli2016 for feedback since it's been a few weeks.

jboeuf commented:
JIangtao, are you taking this one or do you want me to have a look?

On Fri, Feb 23, 2018 at 1:20 PM, David <notifications@github.com> wrote:

> Gentle ping to @jboeuf <https://github.com/jboeuf> @jiangtaoli2016
> <https://github.com/jiangtaoli2016> for feedback since it's been a few
> weeks.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/grpc/grpc/pull/14387#issuecomment-368141665>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AF-P7SrH-iMOwu81QM6CjqwdJ-8asUDiks5tXyuagaJpZM4R_ps->
> .
>

jiangtaoli2016 commented:
I can review. I was waiting for @pmarks-net to approve first.

jiangtaoli2016 commented:
@dcow could you please resolve conflict with master, so that I can trigger some tests?

dcow commented:
@jiangtaoli2016 will do. I'll address @pmarks-net comment about the extraneous allocation at the same time.

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE                                                                                            FILE SIZE
 ++++++++++++++ GROWING                                                                              ++++++++++++++
  [NEW] +8.88Ki src/core/lib/security/transport/security_connector.cc                                +8.88Ki  [NEW]
      [NEW] +1.92Ki [Other]                                                                              +1.92Ki  [NEW]
      [NEW]    +666 grpc_ssl_channel_security_connector_create                                              +666  [NEW]
      [NEW]    +629 try_fetch_ssl_server_credentials                                                        +629  [NEW]
      [NEW]    +537 fake_check_peer(grpc_security_connector*, tsi_peer, grpc_auth_context**, grpc_closur    +537  [NEW]
      [NEW]    +526 fake_channel_check_peer                                                                 +526  [NEW]
      [NEW]    +518 tsi_shallow_peer_from_ssl_auth_context                                                  +518  [NEW]
      [NEW]    +500 grpc_ssl_server_security_connector_create                                               +500  [NEW]
      [NEW]    +479 [Unmapped]                                                                              +479  [NEW]
      [NEW]    +452 compute_default_pem_root_certs_once                                                     +452  [NEW]
      [NEW]    +409 tsi_ssl_peer_to_auth_context                                                            +409  [NEW]
      [NEW]    +404 ssl_check_peer(grpc_security_connector*, char const*, tsi_peer const*, grpc_auth_con    +404  [NEW]
      [NEW]    +343 fake_check_target                                                                       +343  [NEW]
      [NEW]    +223 ssl_host_matches_name                                                                   +223  [NEW]
      [NEW]    +216 grpc_channel_security_connector_cmp                                                     +216  [NEW]
      [NEW]    +202 ssl_channel_check_call_host                                                             +202  [NEW]
      [NEW]    +183 grpc_security_connector_check_peer                                                      +183  [NEW]
      [NEW]    +182 ssl_channel_add_handshakers                                                             +182  [NEW]
      [NEW]    +176 grpc_fake_channel_security_connector_create                                             +176  [NEW]
      [NEW]    +166 grpc_channel_security_connector_check_call_host                                         +166  [NEW]
      [NEW]    +160 ssl_server_add_handshakers                                                              +160  [NEW]
      [NEW]    +154 grpc_server_security_connector_cmp                                                      +154  [NEW]
  +2.0%    +304 src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.cc                          +304  +2.0%
       +28%    +289 grpc_core::(anonymous namespace)::GrpcLb::ProcessChannelArgsLocked                      +289   +28%
      +2.9%     +10 [Unmapped]                                                                               +10  +2.9%
      [NEW]      +5 grpc_core::(anonymous namespace)::BalancerNameCmp                                         +5  [NEW]
  +0.0%     +70 [None]                                                                                   +78  +0.0%
      +0.0%     +30 [Unmapped]                                                                               +70  +0.0%
      +2.9%     +32 [None]                                                                                     0  [ = ]
      [NEW]     +32 grpc_lb_policy_grpclb_build_lb_channel_args(grpc_slice_hash_table*, grpc_core::FakeR     +32  [NEW]
      [NEW]     +16 CSWTCH.75                                                                                +16  [NEW]
      [NEW]      +8 grpc_lb_policy_grpclb_create_lb_channel(char const*, grpc_client_channel_factory*, g      +8  [NEW]

 -------------- SHRINKING                                                                            --------------
  [DEL] -8.82Ki src/core/lib/security/security_connector/security_connector.cc                       -8.82Ki  [DEL]
      [DEL] -1.92Ki [Other]                                                                              -1.92Ki  [DEL]
      [DEL]    -666 grpc_ssl_channel_security_connector_create                                              -666  [DEL]
      [DEL]    -629 try_fetch_ssl_server_credentials                                                        -629  [DEL]
      [DEL]    -537 fake_check_peer(grpc_security_connector*, tsi_peer, grpc_auth_context**, grpc_closur    -537  [DEL]
      [DEL]    -526 fake_channel_check_peer                                                                 -526  [DEL]
      [DEL]    -518 tsi_shallow_peer_from_ssl_auth_context                                                  -518  [DEL]
      [DEL]    -500 grpc_ssl_server_security_connector_create                                               -500  [DEL]
      [DEL]    -492 [Unmapped]                                                                              -492  [DEL]
      [DEL]    -452 compute_default_pem_root_certs_once                                                     -452  [DEL]
      [DEL]    -409 tsi_ssl_peer_to_auth_context                                                            -409  [DEL]
      [DEL]    -404 ssl_check_peer(grpc_security_connector*, char const*, tsi_peer const*, grpc_auth_con    -404  [DEL]
      [DEL]    -343 fake_check_target                                                                       -343  [DEL]
      [DEL]    -216 grpc_channel_security_connector_cmp                                                     -216  [DEL]
      [DEL]    -202 ssl_channel_check_call_host                                                             -202  [DEL]
      [DEL]    -183 grpc_security_connector_check_peer                                                      -183  [DEL]
      [DEL]    -182 ssl_channel_add_handshakers                                                             -182  [DEL]
      [DEL]    -176 grpc_fake_channel_security_connector_create                                             -176  [DEL]
      [DEL]    -166 grpc_channel_security_connector_check_call_host                                         -166  [DEL]
      [DEL]    -160 ssl_server_add_handshakers                                                              -160  [DEL]
      [DEL]    -154 grpc_server_security_connector_cmp                                                      -154  [DEL]
      [DEL]    -146 ssl_host_matches_name                                                                   -146  [DEL]
 -54.5%    -462 src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb_channel_secure.cc           -462 -54.5%
      [DEL]    -815 grpc_lb_policy_grpclb_modify_lb_channel_args                                            -815  [DEL]
     -63.6%     -14 [Unmapped]                                                                               -14 -63.6%
      [DEL]      -5 balancer_name_cmp_fn                                                                      -5  [DEL]
      [DEL]      -5 destroy_balancer_name                                                                     -5  [DEL]
  -9.0%    -240 src/core/ext/transport/chttp2/transport/bin_decoder.cc                                  -240  -9.0%
      [DEL]    -238 grpc_chttp2_base64_infer_length_after_decode                                            -238  [DEL]
     -15.4%      -2 [Unmapped]                                                                                -2 -15.4%

  -0.0%    -264 TOTAL                                                                                   -256  -0.0%


****************************************************************

libgrpc++.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++

 -------------- SHRINK --------------
  [ = ]       0 [None]     -16  -0.0%

  [ = ]       0 TOTAL      -16  -0.0%



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

dcow commented:
@jiangtaoli2016 I rebased to fix the conflicts and addressed @pmarks-net concerns about the allocation (but haven't tested yet).

I did spend a bit of time looking into whether this could viably go in `gpr_split_host_port`. Unfortunately there would be extra work required because the result of `gpr_split_host_port` is used to build the address structures and the expectation is that zone-id is retained. One could entertain modifying `gpr_split_host_port` such that it returns host, zone-id, and port which would mean logic to locate the `%` character could be consolidated and the `sockaddr_in6` creation code cleaned up a tad. But the behavior of security connector comparisons is affected by the presence of a zone-id and host%zone-id would need to be smashed back together when populating target_address (or the old function retained and used), which is annoying.

As a slight side note: I'm not familiar with exactly how the result of comparing security connectors is used functionally throughout grpc, but it strikes me that since port is stripped from the target address when creating a security connector, two connectors pointing at the same host but different ports would compare equal (unless something in their base layers prevents that). However, you generally can't assume one host would have the same security parameters for every service it's serving. If grpc were using these comparisons to make security decisions, this _could_ be a security issue. Or perhaps there could be a latent functional issue when using multiple grpc channels to connect to the same host.

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  [ = ]       0 [None]     +40  +0.0%

  [ = ]       0 TOTAL      +40  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

jiangtaoli2016 commented on src/core/lib/security/security_connector/security_connector.cc:
```diff
@@ -710,13 +710,16 @@ static int ssl_host_matches_name(const tsi_peer* peer, const char* peer_name) {
+  if (!peer_name) return 0;
+
+  // IPv6 zone-id should not be included in comparisons.
+  char *const zone_id = strchr(allocated_name, '%');
```
nit: use `char* const zone_id` to be consistent with grpc coding style.

dcow commented:
@jiangtaoli2016 I'm having a few issues running the tests and code formatter.

### python2.7

    $ pip install six google-api-python-client

When I run `tools/run_tests/run_tests.py` I get:

```
Unexpected error getting flaky tests: Traceback (most recent call last):
  File "tools/run_tests/run_tests.py", line 1421, in <module>
    for test in get_bqtest_data():
  File "tools/run_tests/run_tests.py", line 72, in get_bqtest_data
    import big_query_utils
  File "/Users/dcow/eero/grpc/tools/gcp/utils/big_query_utils.py", line 21, in <module>
    from apiclient import discovery
ImportError: cannot import name discovery

languages with custom make options cannot be built simultaneously with other languages
```
I try:

    $ pip install discovery
    Requirement already satisfied: discovery in ...

### python3    

```
$ tools/run_tests/run_tests.py
  File "tools/run_tests/run_tests.py", line 1680
    except subprocess.CalledProcessError, e:
                                        ^
SyntaxError: invalid syntax
```

### clang-format

```
$ CLANG_FORMAT_ROOT=~/grpc tools/dockerfile/grpc_clang_format/clang_format_all_the_things.sh
xargs: clang-format-5.0: No such file or directory

$ clang-format -version
clang-format version 7.0.0 (tags/google/stable/2018-01-11)

brew install clang-format-5.0
Error: No available formula with the name "clang-format-5.0" 
```

Where can I get `clang-format-5.0`?

dcow commented on test/core/security/security_connector_test.cc:
```diff
@@ -29,6 +29,7 @@
 #include "src/core/lib/gpr/tmpfile.h"
 #include "src/core/lib/security/context/security_context.h"
 #include "src/core/lib/security/security_connector/security_connector.h"
+#include "src/core/lib/security/security_connector/security_connector.cc"
```
There are a few options here regarding `ssl_host_matches_name`:

1. import the source file and test the static function
2. add a proxy to the header in the test section
3. remove static and promote the function itself to the public test api in the header
4. move the logic to `tsi` function

4 seems best but means it's behind the tsi interface now so some future other implementation of the interface would have to also include this logic.. could be a feature. This is also the most expensive option because it incurs an allocation. And, the logic in `tsi_ssl_peer_matches_name` is more fragile.

dcow commented:
@jiangtaoli2016 ping. I'll have time to work on this tomorrow if you have any pointers.

dcow commented:
@jiangtaoli2016 I ran `clang_format_all_the_things.sh` but it only changed files I have not touched.

dcow commented:
@jiangtaoli2016 my bad I ran format_all_the_things on the wrong branch. I've addressed the style issues. I'd love comments on the best approach for the tests since the function we're want to test is static.

jiangtaoli2016 commented on test/core/security/security_connector_test.cc:
```diff
@@ -340,6 +341,29 @@ static grpc_ssl_roots_override_result override_roots_permanent_failure(
+  GPR_ASSERT(tsi_construct_peer(1, &peer) == TSI_OK);
+  for (size_t i = 0; i < GPR_ARRAY_SIZE(addresses); i++) {
+    GPR_ASSERT(tsi_construct_string_peer_property_from_cstring(
+                   TSI_X509_SUBJECT_ALTERNATIVE_NAME_PEER_PROPERTY, san_ips[i],
```
This line seems not formatted properly. Could you please re-run grpc_clang_format?

jiangtaoli2016 commented on test/core/security/security_connector_test.cc:
```diff
@@ -28,6 +28,7 @@
 #include "src/core/lib/gpr/string.h"
 #include "src/core/lib/gpr/tmpfile.h"
 #include "src/core/lib/security/context/security_context.h"
+#include "src/core/lib/security/security_connector/security_connector.cc"
```
Why you need this?

dcow replied:
`ssl_host_matches_name` is `static`. Would love ideas on how to make this better.

dcow replied:
see: https://github.com/grpc/grpc/pull/14387#discussion_r171817241

jiangtaoli2016 replied:
Please don't include .cc file in the header. 

You can make ssl_host_matches_name() a non-static function, and add the function declaration at security_connector.h, on the section of "exposed for testing only".
https://github.com/grpc/grpc/blob/master/src/core/lib/security/security_connector/security_connector.h#L245

In the future, security_connector will be re-written in C++, we will have better way of testing through mock testing class.

dcow replied:
No problem. Sounds good.

jiangtaoli2016 commented on test/core/security/security_connector_test.cc:
```diff
@@ -340,6 +341,29 @@ static grpc_ssl_roots_override_result override_roots_permanent_failure(
+  const char* san_ips[] = {
+      "2001:db8::1",
+      "fe80::abcd:ef65:4321",
+      "fd11:feed:beef:0:cafe::4",
```
Could you please add test cases where san_ip has port and/or zone_id?

e.g. "128.10.0.1:8888", "[2001:db8::1]:8080", "[2001:db8::1%em1]:8080"

dcow replied:
sure!

jiangtaoli2016 commented:
@dcow sorry for delay. A couple of comments.

dcow commented:
@jiangtaoli2016 let me know when you're done reviewing and I will squash.

dcow commented:
@markdroth are you able to ping anyone for me? I'm pretty patient but I also want to get this wrapped up.

jiangtaoli2016 commented on src/core/lib/security/security_connector/security_connector.h:
```diff
@@ -238,6 +238,9 @@ grpc_security_status grpc_ssl_server_security_connector_create(
                                                        const char* name);
 
+/* Hostname verification entry point. */
+int ssl_host_matches_name(const tsi_peer* peer, const char* peer_name);
```
Could you please move this declaration to the section of "exposed for testing only" section?

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  +0.0%     +96 [None]     +32  +0.0%

  +0.0%     +96 TOTAL      +32  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

jiangtaoli2016 commented:
@dcow Somehow this PR was buried in my email box without noticing it. Will merge after you change in the header file and after tests pass.

dcow commented:
@jiangtaoli2016 nit addressed

jiangtaoli2016 commented on src/core/lib/security/security_connector/security_connector.h:
```diff
@@ -244,6 +244,9 @@ tsi_peer tsi_shallow_peer_from_ssl_auth_context(
     const grpc_auth_context* auth_context);
 void tsi_shallow_peer_destruct(tsi_peer* peer);
 
```
sorry I was not clear. Please delete line 246 & 247.

ssl_host_matches_name belongs to functions that are testing only.

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

jiangtaoli2016 commented:
@markdroth Let me know if you are comfortable merging it.

markdroth commented on src/core/lib/security/security_connector/security_connector.h:
```diff
@@ -243,6 +243,7 @@ grpc_auth_context* tsi_ssl_peer_to_auth_context(const tsi_peer* peer);
 tsi_peer tsi_shallow_peer_from_ssl_auth_context(
     const grpc_auth_context* auth_context);
 void tsi_shallow_peer_destruct(tsi_peer* peer);
+int ssl_host_matches_name(const tsi_peer* peer, const char* peer_name);
```
If this function is going to be exposed (even if just for testing), it should have a `grpc_` prefix, to avoid polluting the global symbol namespace.

dcow replied:
what about promoting it to be part of the `tsi`?

markdroth replied:
@jiangtaoli2016's team owns the TSI API, so he can speak to whether he wants this in that namespace.

Note that the existing `tsi_*` functions in this header file are IMHO either incorrectly named or else are in the wrong place.  If they are actually part of TSI, they belong somewhere in `src/core/tsi`; otherwise, they should be named with a `grpc_` prefix instead of a `tsi_` prefix.  But this is a pre-existing problem, not something you should need to fix in this PR.

jiangtaoli2016 replied:
It is pre-existing. 
@dcow how about you change to grpc_ prefix on this particular function, i.e., grpc_ssl_host_matches_name.

I will have a separate PR to clean up the rest.

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

dcow commented:
@markdroth fixed

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  +0.0%     +96 [None]     +88  +0.0%

  +0.0%     +96 TOTAL      +88  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```

markdroth commented:
Looks like you need to run clang-format on security_connector_test.cc.

dcow commented:
@markdroth done.

grpc-testing commented:
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  +0.0%     +96 [None]     +88  +0.0%

  +0.0%     +96 TOTAL      +88  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

grpc-testing commented:
```
[trickle] No significant performance differences
```

grpc-testing commented:
```
[microbenchmarks] No significant performance differences
```
